proxy/envoy/config.go

```diff
@@ -158,62 +162,88 @@ func buildSidecar(env proxy.Environment, sidecar proxy.Node) (Listeners, Cluster
 	services := env.Services()
 	managementPorts := env.ManagementPorts(sidecar.IPAddress)
 
-	inbound, inClusters := buildInboundListeners(env.Mesh, sidecar, instances)
-	outbound, outClusters := buildOutboundListeners(env.Mesh, sidecar, instances, services, env)
-	mgmtListeners, mgmtClusters := buildMgmtPortListeners(env.Mesh, managementPorts, sidecar.IPAddress)
-
-	listeners := append(inbound, outbound...)
-	clusters := append(inClusters, outClusters...)
-
-	// If management listener port and service port are same, bad things happen
-	// when running in kubernetes, as the probes stop responding. So, append
-	// non overlapping listeners only.
-	for i := range mgmtListeners {
-		m := mgmtListeners[i]
-		c := mgmtClusters[i]
-		l := listeners.GetByAddress(m.Address)
-		if l != nil {
-			glog.Warningf("Omitting listener for management address %s (%s) due to collision with service listener %s (%s)",
-				m.Name, m.Address, l.Name, l.Address)
-			continue
+	listeners := make(Listeners, 0)
+	clusters := make(Clusters, 0)
+
+	if env.Mesh.ProxyListenPort > 0 {
+		inbound, inClusters := buildInboundListeners(env.Mesh, sidecar, instances)
+		outbound, outClusters := buildOutboundListeners(env.Mesh, sidecar, instances, services, env.IstioConfigStore)
+		mgmtListeners, mgmtClusters := buildMgmtPortListeners(env.Mesh, managementPorts, sidecar.IPAddress)
+
+		listeners = append(listeners, inbound...)
+		listeners = append(listeners, outbound...)
+		clusters = append(clusters, inClusters...)
+		clusters = append(clusters, outClusters...)
+
+		// If management listener port and service port are same, bad things happen
+		// when running in kubernetes, as the probes stop responding. So, append
+		// non overlapping listeners only.
+		for i := range mgmtListeners {
+			m := mgmtListeners[i]
+			c := mgmtClusters[i]
+			l := listeners.GetByAddress(m.Address)
+			if l != nil {
+				glog.Warningf("Omitting listener for management address %s (%s) due to collision with service listener %s (%s)",
+					m.Name, m.Address, l.Name, l.Address)
+				continue
+			}
+			listeners = append(listeners, m)
+			clusters = append(clusters, c)
+		}
+
+		// set bind to port values for port redirection
+		for _, listener := range listeners {
+			listener.BindToPort = false
 		}
-		listeners = append(listeners, m)
-		clusters = append(clusters, c)
-	}
 
-	// set bind to port values for port redirection
-	for _, listener := range listeners {
-		listener.BindToPort = false
+		// add an extra listener that binds to the port that is the recipient of the iptables redirect
+		listeners = append(listeners, &Listener{
+			Name:           VirtualListenerName,
+			Address:        fmt.Sprintf("tcp://%s:%d", WildcardAddress, env.Mesh.ProxyListenPort),
```

Comment:
it's definitely needed for the virtual listener to redirect to another listener.